### ISSUE TITLE
=children/=focus/Zoom

### ISSUE THREAD

**Issue by raineorshine** [cybersemics/em#651]
Allow `=focus/Zoom` to be added to the `=children` attribute. This should create the same effect as adding `=focus/Zoom` to each child.

**Comment by pushkov-fedor** [comment]
Raine, hi! Could you please explain in more detail what` =focus/Zoom` attribute does and what difference between `Normal `and `Zoom`? I have read Technical Overview, what does **additional focus** mean?'
I have such tree:
- a
  - b
    - =focus/Zoom
  - sibling

and i haven't seen difference with and without this attribute. Also could you point me out please where i can read about `=children` attribute. I didn't find anything about it in Technical Overview. Thanks!

**Comment by raineorshine** [comment]
Sure! `=focus/Zoom` is a special mode that hides the parent and siblings of a given thought when the cursor is on it or on a descendant in order to provide more focus. "Additional focus" just means fewer distractions for the user. It is not a technical term.

I write `=focus/Zoom` as shorthand for a `Path` like `[..., '=focus', 'Zoom']`. So the tree would look like:

```
- a
  - b
    - =focus
      - Zoom
  - c
```

> Also could you point me out please where i can read about `=children` attribute. I didn't find anything about it in Technical Overview. Thanks!

Yes, the purpose of the `=children` attribute is to apply arbitrary attributes to all children. It currently only works with a couple attributes like `=style` and `=bullet`.

e.g. This would make `b` and `c` the color `tomato`:

```
- a
  - =children
    - =style
      - color
        - tomato
  - b
  - c
```

I will add this to the Wiki!

**Comment by pushkov-fedor** [comment]
I see now, thanks! As i understand we just need to add this attribute to all children. I couldn't figure out, what attribute function is uses and where applying attributes to a Thought happens. Could you point me out please where to find it?

**Comment by raineorshine** [comment]
Take a look at how `styleChildren` is passed from `Subthoughts` to `Thought`. `=children/=focus` should work exactly the same way.

https://github.com/cybersemics/em/blob/dev/src/components/Subthoughts.tsx#L553-L557

As a fun challenge, you could abstract the solution so that *any* attribute could be applied via `=children`.

**Comment by pushkov-fedor** [comment]
Yes, thanks, i see! But i was talking about about all attributes in general. As i understand, we don't store such thought attributes somewhere directly, instead we calculate it from context just checking if there thoughts exists with certain metaprogramming value. Am i right?

**Comment by raineorshine** [comment]
> Yes, thanks, i see! But i was talking about about all attributes in general. 

Attributes are applied ad hoc. That is, there is no generalized attribute evaluation. Each specific attribute is checked for when it may affect the layout/rendering, e.g. `=style` gets read when the `style` property is rendered, and `=readonly` gets read when the `onChange` handler is called.
 
> As i understand, we don't store such thought attributes somewhere directly, instead we calculate it from context just checking if there thoughts exists with certain metaprogramming value. Am i right?

Yes, if I am understanding you correctly, metaprogramming attributes are just thoughts, stored like any other thought. They are intrinsic and reflexive rather than extrinsic. You can modify the behavior of thoughts with thoughts.

**Comment by raineorshine** [comment]
Fixed in d12d416e
